myia/vm.py

```diff
@@ -0,0 +1,222 @@
+            """Build a closure."""
+            self.graph = graph
+            self.frame = frame
+
+    class Jump(ANFNode):
+        """Jump Node for out of scope values."""
+
+        def __init__(self, frame: 'VMFrame') -> None:
+            """Set a jump."""
+            super().__init__([], frame, None)
+
+    def __init__(self, graph: Graph, args: List[Any], parent: 'VMFrame',
+                 closure: 'VMFrame' = None) -> None:
+        """Build a frame."""
+        self.graph = graph
+        self.args = args
+        self.parent = parent
+        self.closure = closure
+        self.values: Dict[ANFNode, Any] = dict()
+        self.todo: List[ANFNode] = [self.graph.return_]
+
+    def run(self) -> 'VMFrame':
+        """Run the graph to completion and return the result."""
+        while True:
+            try:
+                res = self.advance()
+            except StopIteration:
+                # Return is handled here because the flow control is a bit
+                # too complex for the parent frame to get a value back
+                # from the called frame.
+                if self.parent is not None:
+                    self.parent.values[self.parent.todo[-1]] = self.result()
+                    self.parent.todo.pop()
+                return self.parent
+            if res is not None:
+                return res
+
+    def tail(self) -> bool:
+        """Are we in a tail call?."""
+        return self.todo[-2] is self.graph.return_
+
+    def done(self) -> bool:
+        """Are we done?."""
+        return len(self.todo) == 0
+
+    def result(self):
+        """Get this frame's result."""
+        return self.values[self.graph.return_]
+
+    def get_value(self, node: ANFNode):
+        """Get the value for a node.
+
+        This will check if we've already computed a value for it and
+        add it to the queue otherwise. If there was no value, we stop
+        the evaluation of the current node and restart the step, since
+        no useful work was done.
+
+        If the value we want is in another graph, we will resume
+        execution in a different frame and jump back to this one later.
+        """
+        frame = self
+        while frame and node.graph and node.graph is not frame.graph:
+            frame = frame.closure
+        if node not in frame.values:
+            # we set up a fake node to jump back to this frame after
+            # the value that we need has been evaluated.
+            if frame is not self:
+                frame.todo.append(self.Jump(self))
+            frame.todo.append(node)
+            if frame is not self:
+                raise self.Return(frame)
+            else:
+                raise self.Continue()
+        return frame.values[node]
+
+    def wrap_closure(self, value):
+        """Wrap graphs that are closures."""
+        if isinstance(value, Graph):
+            if any(n.graph and n.graph is not value
+                   for n in dfs(value.return_)):
```

Comment:
Pretty sure this is causing the issue with https://github.com/mila-udem/myia/pull/8#issuecomment-358799988. In short, this will miss free variables: if this is graph `G`, another graph `H` is nested inside `G`, and `H` accesses a variable that's outside the scope of both `G` and `H`, then that variable is a free variable of both graphs. But you're not looking inside the graphs encountered in the `dfs`, so you will miss it.

I think that for this interpreter it's best not to try to be clever: create a closure all the time.

Reply:
If we don't have a reliable way to detect closures, this is the safest choice, yes.  However I fear it will keep all the stacks around for loops and cost a lot of memory.

Reply:
It's not the stack per se it's supposed to save, just the environment, so a closure should never keep more "frames" alive than the lexical nesting depth. I think it'd be safe to set a frame's parent to `None` once it returns, so that the GC can reclaim it.

Reply:
Or maybe it'd be better to split functionality into `Environment` (containing `values` and a parent `environment`) and `Frame` (containing a `return_node`, an `environment` and a `todo`), and have closures store the environment only. I think it'd be a bit clearer, and then you can safely reuse `Frame` objects in tail calls and whatnot.

Reply:
But if I understand correctly, `parent` is the caller frame. The caller frame is not part of the environment. What it represents essentially is "what are we doing once we've computed the `return_` node?" If we're missing some nodes and we need to compute them later, we're doing something different afterwards: we're not returning to the parent twice.

So, the way I think it should work is, suppose you have:

```python
def f(x):
    a = x + x
    def g():
        return a
    return g
```

When `f` is called, we create `Frame(f.env, f.return_)`. When `g` is encountered, we create `Closure(g, new_env(f.env))`. Now, when that closure is called and it sees it needs `a`, and it appears `a` has not been computed, then it creates a *new frame*, `Frame(f.env, f.a)`. That frame uses `f.env` so it has access to everything previously computed, but the "return node" is a different one. In other words, you have one environment for `f`, but one frame for every time you need to compute something there.

Reply:
> The more I think about it, the more I realize that we probably need to do a pass to mark which graphs are closures and up to where do we need to preserve the environment to capture everything they need.

Also, just by the way, a precise account of free variables *is* needed for `grad`, so I am working on this right now. I don't think there's any need for this in the VM, but in either case, don't do it.

Reply:
> But if I understand correctly, parent is the caller frame. The caller frame is not part of the environment. What it represents essentially is "what are we doing once we've computed the return_ node?" If we're missing some nodes and we need to compute them later, we're doing something different afterwards: we're not returning to the parent twice.

Yes, that is mostly correct.  But the "missing" nodes could need to be computed in the middle of a nested call.  They could also be required to compute the return value later.  So we need to ensure that all the nodes that are evaluated for a specific application of a graph happen in the same environment.

> When f is called, we create Frame(new_env(f.env), f.return_). When g is encountered, we create Closure(g, f.env). Now, when that closure is called and it sees it needs a, and it appears a has not been computed, then it creates a new frame, Frame(f.env, f.a) (no call to new_env here). That frame uses f.env so it has access to everything previously computed, but the "return node" is a different one. In other words, you have one environment for f, but one frame for every time you need to compute something there.

What is `f.env` in your example?  As far as I know, `f` is a Graph and those don't have an env.  And even if they did it couldn't be the basis for the frame environment since that one needs the parameters for a specific application and the mapping of the values for the nodes that might depend on the values of the parameters.

Reply:
Sorry, I was being loose with notation. You get the env from the frame when you create a closure, and from the closure when you call it. I mean something like this:

```python
def call(fn, args):
    graph = fn.graph if isinstance(fn, Closure) else fn
    parent_env = fn.env if isinstance(fn, Closure) else None
    env = Environment(parent_env=parent_env, graph=graph)
    for param, arg in zip(graph.parameters, args):
        env[param] = arg
    frame = Frame(env=env, ret=graph.return_)
    exec_frame(frame)

def wrap_closure(current_frame, graph):
    return Closure(current_frame.env, graph)

def eval_node(current_frame, node):
    if node.graph is not current_frame.graph:
        env = current_frame.env
        while env.graph is not node.graph:
            env = env.parent_env
        frame = Frame(env=env, ret=node)
        exec_frame(frame)
    ...
```

Is that any clearer? It's a bit difficult to explain, but I think the idea of saving "frames" muddied the waters. It's a red herring, it's just environments that you need to save.

Reply:
I see what you are getting at.  I'll try to rework the code to follow that pattern.

Reply:
Actually, I thought about it more, and I think this won't work, but for a different reason than I thought. Basically, consider:

```python
def f(x):
    def g(z):
        return h(z + x)
    def h(z):
        return g(z * x)
    return g(3)
```

This doesn't terminate, but that's besides the point -- the calls are tail calls, so this should loop forever without growing memory.

However, when you call `g` and you encounter `h`, the code I gave will naively give it `g` as its environment, so it'll have environment chain `[f, g]`. This is wrong, of course, because `h` is not nested in `g`, but this isn't written anywhere in the representation. Nonetheless, when `h` is called, it will see `g`, and it will naturally assume that `g` is nested in `h`, so it will create the environment `[f, g, h]`. Then `[f, g, h, g]`, and so on. The stack won't grow, but the chain of environments will, because both functions think they are nested in each other.

Maybe it'd kinda-work if the closures are cached, but honestly I think we should forget about evaluating this representation directly and preprocess with closure conversion. Then it'll be way more straightforward. I'm already working on that because of `grad`, it's a bit more complicated I thought it'd be, but I think it should be done by next meeting.

Reply:
Yes, it would be much easier to work on a closure-converted version where everything is passed around explicitly.  If it is required for grad to work properly, then we *can* also make it a requirement for the vm, but I would prefer if it wasn't.

We can talk about it before the next meeting, yes.

Reply:
If possible, I agree it would be nice not to have to rely on closure conversion (debugging closure conversion and the parser would be easier that way).